contracts/rust/adapter/src/stake_table.rs

```diff
@@ -24,10 +59,120 @@ impl From<EdOnBN254PointSol> for StateVerKey {
     }
 }
 
+pub fn sign_address_bls(bls_key_pair: &BLSKeyPair, address: Address) -> G1PointSol {
+    bls_key_pair
+        .sign(&address.abi_encode(), CS_ID_BLS_BN254)
+        .sigma
+        .into_affine()
+        .into()
+}
+
+pub fn sign_address_schnorr(schnorr_key_pair: &StateKeyPair, address: Address) -> Bytes {
+    let msg = [hash_bytes_to_field(&address.abi_encode()).expect("hash to field works")];
+    let mut buf = vec![];
+    schnorr_key_pair
+        .sign(&msg, CS_ID_SCHNORR)
+        .serialize_compressed(&mut buf)
+        .expect("serialize works");
+    buf.into()
+}
+
+fn authenticate_schnorr_sig(
+    schnorr_vk: &StateVerKey,
+    address: Address,
+    schnorr_sig: &[u8],
+) -> Result<(), StakeTableSolError> {
+    let msg = [hash_bytes_to_field(&address.abi_encode()).expect("hash to field works")];
+    let sig = schnorr::Signature::<EdwardsConfig>::deserialize_compressed(schnorr_sig)?;
+    schnorr_vk.verify(&msg, &sig, CS_ID_SCHNORR)?;
+    Ok(())
+}
+
+fn authenticate_stake_table_validator_event(
+    account: Address,
+    bls_vk: G2PointSol,
+    schnorr_vk: EdOnBN254PointSol,
+    bls_sig: G1PointSol,
+    schnorr_sig: &[u8],
+) -> Result<(), StakeTableSolError> {
+    // TODO(alex): simplify this once jellyfish has `VerKey::from_affine()`
+    let bls_vk = {
+        let bls_vk_inner: ark_bn254::G2Affine = bls_vk.into();
+        let bls_vk_inner = bls_vk_inner.into_group();
+
+        // the two unwrap are safe since it's BLSPubKey is just a wrapper around G2Projective
+        let mut ser_bytes: Vec<u8> = Vec::new();
+        bls_vk_inner.serialize_uncompressed(&mut ser_bytes).unwrap();
+        BLSPubKey::deserialize_uncompressed(&ser_bytes[..]).unwrap()
+    };
+    let msg = account.abi_encode();
+    let sig = {
+        let sigma_affine: ark_bn254::G1Affine = bls_sig.into();
+        BLSSignature {
+            sigma: sigma_affine.into_group(),
+        }
+    };
+    if !bls_vk.validate(&sig, &msg) {
```

Comment:
I might have missed it but do we have a test for bls vk validation against its sigs?

Reply:
You mean unittests? I didn't add them because @alxiong wrote it, so it should be correct :joy: added some unittests here: 985ce8d18874437d80d2fae2014f3ef84dbc35bc